aip/general/0250.md

```diff
@@ -0,0 +1,107 @@
+---
+
+# Short names
+
+System-generated identifiers have many benefits, but are not easy for humans
+use.
+
+## Guidance
+
+Resources with system-generated identifiers may choose to support a field
+`short_name` of type string, which **must** refer to a terse, human-friendly
+value which uniquely identifies a resource of a specific type within a
+namespace.
+
+* `short_name` **must** have clear documentation for its namespace.
+
+* `short_name` **must** support at least the character set of
+  `/^[a-z][a-z0-9.-]+[a-z0-9]$/` (lowercase
+  [RFC-1034](https://tools.ietf.org/html/rfc1034)).
+
+* `short_name` **must** have a maximum length of no less than 30 characters.
+
+* `short_name` **should** be immutable, which is safer for references in
+  configuration files.
+
+### Usage
+
+To illustrate this concept, let us imagine a bookstore that uses a
+config-as-code system to manage which books are to be on sale each week.
+
+```proto
+message Book {
+  // The resource name of the book.
+  // Example: publishers/1235/books/6789
+  string name = 1;
+
+  // A short identifier for human use in the namespace of the publisher.
+  // Example: mary-poppins
+  string short_name = 2;
+
+  // The title of the book.
+  // Example: "Mary Poppins"
+  sting title = 3;
+}
+```
+
+```yaml
+- week_start: 2020-09-13
+  books_on_sale:
+    - book: publishers/1235/books/6789
+      price: $5.99
+    - book: publishers/2342/books/9827
+      price: $25.99
+```
+
+While this configuration is clear to machines, the human reader is at a loss for
+what is on sale. Tooling can choose to support `short_name` as follows:
+
+```yaml
+- week_start: 2020-09-13
+  books_on_sale:
+    - book: 1235/mary-poppins
+      price: $5.99
+    - book: 2342/thing-explainer
+      price: $25.99
+```
+
+Now the configuration is much more understandable to humans. The tooling can
+convert "1235/mary-poppins" to the resource name "publishers/1235/books/6789" by
+using the list method of the Books API (eg,.
+`ListBooks(parent="publishers/2345", filter="short_name:mary-poppins")`).
+
+Use of the `short_name` may also be appropriate in reporting scenarios, where
+tooling can find strings like "publishers/1235/books/6789" and replace them
+with "1235/mary-poppins" in contexts where a human is reading the output.
+
+### Context & Namespace
+
+`short_name` requires tools understand the context of the reference and the
+namespace in which the `short_name` is unique. The namespace **should** be based
+on an immutable property of the resource (eg,. its parent or an unchangeable
+portion of its ancestry).
+
+In the example above, the context is a reference to a Book, and the namespace
+is the publisher identifier.
+
+The namespace itself may utilize short names. For example, a database table in a
+project could be referenced as `projects/123/databases/456/table/789` or
+`123/wholesaler-db/books`.
+
+### Use in API calls
+
+Some APIs may choose to allow `short_name` to be used in place of the
```

Comment:
Services **should not** allow `short_name`...

Reply:
done.